raster-foundry/raster-foundry
# Kickoff project layer overview generation reactively

## Overview

This PR enables kicking off project layer overview generation in the below cases:
- adding scenes to a layer
- setting scene manual order
- deleting scenes from a layer
- a scene's (when it is in a project layer) ingest status is updated 
- splitting a layer into multiple ones (adding scenes to new layers -- so the same as the first case under the hood actually)

It removes the generated overview from s3 if there is any in the below cases:
- deleting a layer
- deleting a project
- deleting all scenes from a layer (it also removed the URL from the overview location field from layers table)


### Checklist

- ~Styleguide updated, if necessary~
- ~Swagger specification updated~
- ~Symlinks from new migrations present or corrected for any new migrations~
- ~Any content changes are properly templated using `BUILDCONFIG.APP_NAME`~

## Testing Instructions

- Reassemble the api jar. Spin up frontend and server.
- Create a new project. Look at api server logs. Add a scene to this project (default layer). There should be a message like `Kicking off layer overview creation for project-<project id>-layer-<layer id>`. And it will tell you that it will not invoke lambda function to generate overview because it is not production or staging. But it provides you with the lambda function name it will invoke and the json payload to be sent.
- Test that you see similar logs as the above if you do the following:
    * Add another scene to this project layer
    * Manually reorder these two scenes
    * In V2 UI, split this layer
    * Remove one of the scenes
    * Now pause
- Go to DB, update this layer's `overviews_location` field with a random string. Get the ID of the remaining scene. Update this scenes's ingest status to something other than `INGESTED`, and make yourself the owner of the scene.
- `PUT` to `api/scenes/<scene ID from above>` with the updated `INGESTED` status. The similar log message should show up again
- Remove this layer's last scene on frontend. The logs should tell you `No scenes left in project layer <layer id> with overview location <the url you just updated>`, but it cannot find the resource on S3 (since it is a random string anyways -- this proves that it checks its existence before deleting, and if it finds anything, it will delete it, since now there are no scenes in this layer). Also check in the DB that this layer's `overviews_location` field is now null.
- Add another scene to this layer. Update this layer's `overviews_location` field again. Now go to frontend and delete this layer. The similar deletion message should show up again.
- Create two other layers. Add scenes to these two layers. Update these two layer's `overviews_location` fields. Delete this **project**, two deletion messages should show up in the logs.
- Now push this branch up as a test branch. When it gets deployed, go to staging, and test the above functionalities. See logs in papertrail that lambda functions are called and json payloads are sent. Go to lambda logs, and watch. Go to S3 to check if the resources are there when overview generations are done. Perform similar remove actions as mentioned above and check if they behave as expected.
